torch/csrc/distributed/rpc/utils.h

```diff
@@ -142,50 +127,37 @@ struct TORCH_API LazyStreamContext {
 
   // get a stream for the given device. If it is the first time using that
   // device, allocate a new stream and store it in the map.
-  c10::Stream getStream(c10::DeviceIndex index) {
-    if (!stream_creator_) {
-      throw std::runtime_error(c10::str(
-          "Attempting to access device stream of device ",
-          index,
-          ", but the device doesn't support streams"));
-    }
-    auto iter = streams_.find(index);
+  c10::Stream getStream(c10::Device device) {
+    auto iter = streams_.find(device);
     if (iter == streams_.end()) {
-      auto stream = stream_creator_(index);
-      streams_.emplace(index, stream);
+      auto stream = impl_.getStreamFromPool(device);
+      streams_.emplace(device, stream);
       return stream;
     } else {
       return iter->second;
     }
   }
 
-  std::set<c10::DeviceIndex> devices() const {
-    std::set<c10::DeviceIndex> devices;
+  std::unordered_set<c10::Device> devices() const {
+    std::unordered_set<c10::Device> devices;
     for (const auto& entry : streams_) {
       devices.insert(entry.first);
     }
     return devices;
   }
 
   c10::DeviceType deviceType() const {
-    return device_type_;
+    return impl_.type();
   }
 
  private:
-  std::unordered_map<c10::DeviceIndex, c10::Stream> streams_;
-  c10::DeviceType device_type_;
-  stream_factory_t stream_creator_;
-  stream_factory_t current_stream_provider_;
+  const c10::impl::VirtualGuardImpl impl_;
+  std::unordered_map<c10::Device, c10::Stream> streams_;
 };
 
 inline std::shared_ptr<LazyStreamContext> createLazyStreamContext(
```

Comment:
do we still need this helper? looks like we can directly call the `LazyStreamContext` ctor?